filebeat/input/winlog/input.go

```diff
@@ -105,41 +96,64 @@ func (eventlogRunner) Run(
 	})
 	defer cancelFn()
 
-	// read loop
-	for cancelCtx.Err() == nil {
-		records, err := api.Read()
-		switch err {
-		case nil:
-			break
-		case io.EOF:
-			log.Debugf("End of Winlog event stream reached: %v", err)
+runLoop:
+	for {
+		if cancelCtx.Err() != nil {
 			return nil
-		default:
-			// only log error if we are not shutting down
-			if cancelCtx.Err() != nil {
-				return nil
-			}
-
-			log.Errorf("Error occured while reading from Windows Event Log '%v': %v", source.Name(), err)
-			return err
 		}
 
-		if len(records) == 0 {
-			timed.Wait(cancelCtx, time.Second)
+		evtCheckpoint := initCheckpoint(log, cursor)
+		openErr := api.Open(evtCheckpoint)
+		if eventlog.IsRecoverable(openErr) {
+			log.Errorf("Encountered recoverable error when opening Windows Event Log: %v", openErr)
+			time.Sleep(time.Second * 5)
```

Comment:
I think using `timed.Wait(cancelCtx, 5 * time.Second)` would be better because it would allow the sleep to unblock if cancellation occurs.